resources/analysisTools/bam2fastq/bam2Fastq.sh

```diff
@@ -7,96 +7,265 @@
+                ## can be moved, though.
+                mv "$srcName" "$tgtName" || throw 10 "Could not move file '$srcName' to '$tgtName'"
+            else
+                ## Furthermore, for read groups in the SAM header without reads, no FASTQ is produced. Such read group produce problems
+                ## downstream. So here we create an empty dummy FASTQ. Note that this may include the 'default' read group that Roddy adds
+                ## See "bamListReadGroups.sh".
+                cat /dev/null | gzip -c - > "$tgtName"
+            fi
+        done
+    done
+}
+
+
+processPairedEndWithoutReadGroupsPicard() {
+    local baseName=$(basename "$FILENAME_BAM" .bam)
+
+    ## Extract the output directory from the first lane file path entered.
+    local outputDir="$outputAnalysisBaseDirectory/${baseName}_temp"
+    registerTmpFile "$outputDir"
+
+    ## Write just 2-3 FASTQs, depending on whether unpairedFastq is true.
+    ## We need to add the suffix here such that Picard automatically does the compression.
+    local fastq1BaseName="${baseName}_r1.${FASTQ_SUFFIX}"
+    local tmpFastq1=$(makeTempName "$outputDir" "$fastq1BaseName")
+    local fastq2BaseName="${baseName}_r2.${FASTQ_SUFFIX}"
+    local tmpFastq2=$(makeTempName "$outputDir" "$fastq2BaseName")
+    local PICARD_OPTIONS="$PICARD_OPTIONS COMPRESS_OUTPUTS_PER_RG=$compressIntermediateFastqs FASTQ=$tmpFastq1 SECOND_END_FASTQ=$tmpFastq2"
+    if [[ "${writeUnpairedFastq:-false}" == true ]]; then
+        local fastq3BaseName="${baseName}_r3.${FASTQ_SUFFIX}"
+        local tmpFastq3=$(makeTempName "$outputDir" "$fastq3BaseName")
+        local PICARD_OPTIONS="$PICARD_OPTIONS UNPAIRED_FASTQ=$tmpFastq3"
+    fi
+
+    ## Only process the non-supplementary reads (-F 0x800). BWA flags all alternative alignments as supplementary while the full-length
+    ## reads are exactly the ones not flagged supplementary.
+    local JAVA_OPTIONS="${JAVA_OPTIONS:-$JAVA_OPTS}"
+    "$SAMTOOLS_BINARY" view -u $(samtoolsExclusions) "$FILENAME_BAM" \
+        | "$PICARD_BINARY" $JAVA_OPTIONS SamToFastq $PICARD_OPTIONS INPUT=/dev/stdin
+
+    ## Now make sure that the output files of Picard are renamed correctly.
+    mv "$tmpFastq1" "${FILENAME_UNSORTED_FASTQ[0]}" || throw 10 "Could not move file '$tmpFastq1' to '${FILENAME_UNSORTED_FASTQ[0]}'"
+    mv "$tmpFastq2" "${FILENAME_UNSORTED_FASTQ[1]}" || throw 10 "Could not move file '$tmpFastq2' to '${FILENAME_UNSORTED_FASTQ[1]}'"
+    if [[ "${writeUnpairedFastq:-false}" == true ]]; then
+        mv "$tmpFastq3" "${FILENAME_UNSORTED_FASTQ[2]}" || throw 10 "Could not move file '$tmpFastq3' to '${FILENAME_UNSORTED_FASTQ[2]}'"
+    fi
+}
+
+processSingleEndWithReadGroupsPicard() {
+    throw 1 "processSingleEndWithReadGroups not implemented"
+}
+
+processSingleEndWithoutReadGroupsPicard() {
+    throw 1 "processSingleEndWithoutReadGroups not implemented"
+}
+
+
+
+main() {
+    setUp_BashSucksVersion
```

Comment:
Something more explanation :D

Reply:
Referencing the function definition in the library here. Again explaining everything here is overkill.